DllImportGenerator/DllImportGenerator/DllImportGenerator.cs

```diff
@@ -1,164 +1,183 @@
-                    if (attr.AttributeClass is null)
+                    if (!data.Left.isSupported && data.Right.Any())
                     {
-                        continue;
+                        // We don't block source generation when the TFM is unsupported.
+                        // This allows a user to copy generated source and use it as a starting point
+                        // for manual marshalling if desired.
+                        context.ReportDiagnostic(
+                            Diagnostic.Create(
+                                GeneratorDiagnostics.TargetFrameworkNotSupported,
+                                Location.None,
+                                MinimumSupportedFrameworkVersion.ToString(2)));
                     }
-                    else if (attr.AttributeClass.ToDisplayString() == TypeNames.GeneratedDllImportAttribute)
+                });
+
+            var stubEnvironment = compilationAndTargetFramework
+                .Combine(context.AnalyzerConfigOptionsProvider)
+                .Select(
+                    static (data, ct) =>
+                        new StubEnvironment(
+                            data.Left.compilation,
+                            data.Left.isSupported,
+                            data.Left.targetFrameworkVersion,
+                            data.Right.GlobalOptions,
+                            data.Left.compilation.SourceModule.GetAttributes().Any(attr => attr.AttributeClass?.ToDisplayString() == TypeNames.System_Runtime_CompilerServices_SkipLocalsInitAttribute))
+                );
+
+            var methodSourceAndDiagnostics = methodsToGenerate
+                .Combine(stubEnvironment)
+                .Select(static (data, ct) => new
+                {
+                    data.Left.Syntax,
+                    data.Left.Symbol,
+                    Environment = data.Right
+                })
+                .Select(
+                    (data, ct) =>
                     {
-                        generatedDllImportAttr = attr;
+                        IncrementalTracker?.RecordExecutedStep(new IncrementalityTracker.ExecutedStepInfo(IncrementalityTracker.StepName.CalculateStubInformation, data));
+                        return (data.Syntax, StubContext: CalculateStubInformation(data.Syntax, data.Symbol, data.Environment, ct));
                     }
-                    else if (lcidConversionAttrType is not null && SymbolEqualityComparer.Default.Equals(attr.AttributeClass, lcidConversionAttrType))
+                )
+                .WithComparer(Comparers.CalculatedContextWithSyntax)
+                .Combine(context.AnalyzerConfigOptionsProvider)
+                .Select(
+                    (data, ct) =>
                     {
-                        lcidConversionAttr = attr;
+                        IncrementalTracker?.RecordExecutedStep(new IncrementalityTracker.ExecutedStepInfo(IncrementalityTracker.StepName.GenerateSingleStub, data));
```

Comment:
I don't see why the `IncrementalTracker` should ever be `null`. Is there a case where we don't need tracking? If so, I would like to see a default/nop implementation.

Reply:
Okay. Please ensure that instance/member is documented as test only.